src/metabase/api/notify.clj

```diff
@@ -33,19 +34,64 @@
         db-sync-fn    (if schema? sync-metadata/sync-db-metadata! sync/sync-database!)]
     (api/let-404 [database (t2/select-one Database :id id)]
-      (cond-> (cond
-                table_id   (api/let-404 [table (t2/select-one Table :db_id id, :id (int table_id))]
-                             (future (table-sync-fn table)))
-                table_name (api/let-404 [table (t2/select-one Table :db_id id, :name table_name)]
-                             (future (table-sync-fn table)))
-                :else      (future (db-sync-fn database)))
-        synchronous? deref)))
+      (let [table (cond
+                    table_id   (api/check-404 (t2/select-one Table :db_id id, :id (int table_id)))
+                    table_name (api/check-404 (t2/select-one Table :db_id id, :name table_name)))]
+        (cond-> (future (if table
+                          (table-sync-fn table)
+                          (db-sync-fn database)))
+          synchronous? deref))))
   {:success true})
 
 (defn- without-stacktrace [^Throwable throwable]
   (doto throwable
     (.setStackTrace (make-array StackTraceElement 0))))
 
+(defn- find-and-sync-new-table
+  [database table-name schema-name]
+  (let [driver (driver.u/database->driver database)
+        {db-tables :tables} (driver/describe-database driver database)]
+    (if-let [table (some (fn [table-in-db]
+                           (when (and (= schema-name (:schema table-in-db))
+                                      (= table-name (:name table-in-db)))
+                             table-in-db))
+                         db-tables)]
+      (let [created (sync-tables/create-or-reactivate-table! database table)]
+        (doto created
+          sync/sync-table!
+          sync-util/set-initial-table-sync-complete!))
+      (throw (without-stacktrace
+              (ex-info (trs "Unable to identify table ''{0}.{1}''"
+                            schema-name table-name)
+                       {:status-code 404
+                        :schema_name schema-name
+                        :table_name  table-name}))))))
+
+(api/defendpoint POST "/db/attached_datawarehouse"
+  "Sync the attached datawarehouse. Can provide in the body:
+  - sync_db: <boolean>: if this is true, schema sync the whole attached datawarehouse
+  - table_name and schema_name: both strings. Will look for an existing table and sync it, otherwise will try to find a
+    new table with that name and sync it. If it cannot find a table it will throw an error.
+  - synchronous?: is a boolean value to indicate if this should block on the result."
+  [:as {{:keys [table_name schema_name sync_db synchronous?]} :body}]
+  {table_name   [:maybe ms/NonBlankString]
+   schema_name  [:maybe string?]
+   sync_db      [:maybe ms/BooleanValue]
+   synchronous? [:maybe ms/BooleanValue]}
+  (when (and (not sync_db) (str/blank? table_name))
+    (throw (ex-info (trs "Must provide `sync_db` or a `table_name` to sync") {:status-code 400})))
```

Comment:
Could just use the absence of a `table_name` to indicate a full DB sync instead of a separate flag? (Just an idea, I'm indifferent)

Reply:
Good idea. I've updated it!